### PR TITLE
Feature/dock 2105/improve dockstore yml error messages

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR improves the error message that is written to the github apps logs upon a failure to read/validate `.dockstore.yml`.  This PR implements the portion of https://ucsc-cgl.atlassian.net/browse/DOCK-2105 that should I believe should be hotfixed to 1.12.x.  Another [PR](https://github.com/dockstore/dockstore/pull/4844) implements the portion that might be better merged with develop.

The primary improvements are to include a phrase that indicates that the problem was in `.dockstore.yml`, and prefix the constraint violation message with the property name and append the errant value (if available, abbreviated if necessary).  The user sees it when they expand the log entry:

Screen Shot 2022-04-07 at 10 04 48 PM

Constraint violation messages are/should be normalized to verb phrases, so they can be automatically prefixed with a subject.  The `NotNull` message was changed to a more understandable phrase.

This PR also:
* adds the required annotations to validate the full Dockstore YAML object hierarchy, parts of which were not being validated previously.
* fixes the "at least one workflow, tool, or service" validator so it works, previously it would NPE on empty `workflows` or `tools`. I believe the SonarCloud warnings relating to this fix are erroneous.
* bumps the version to `1.12.1-SNAPSHOT`

Now that the system provides the user with actionable feedback on validation failure, we should consider more extensive validations (we could validate email addresses and orcid ids, check that fields are not blank, etc).

**Issue**
https://ucsc-cgl.atlassian.net/browse/DOCK-2105

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] Check the Snyk dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

### CODE REVIEW

**Review by denis-yuen — COMMENTED** [review]
these two PRs do not have tests, any thoughts on how reasonable they are to test?

**Comment by svonworl** [comment]
> these two PRs do not have tests, any thoughts on how reasonable they are to test?

I can add a few tests, the other PR definitely needs some.

**Review by coverbeck — COMMENTED** [review]
Looks good, but I think there should be some tests.

There are unit tests io.dockstore.common.yaml.DockstoreYamlTest you can update (a lot easier to deal with than integration tests, IMHO).

File: dockstore-common/src/main/java/io/dockstore/common/yaml/Dockstore12Validator.java [hunk 1]
```diff
@@ -29,6 +29,8 @@ public void initialize(final ValidDockstore12 constraintAnnotation) {
 
     @Override
     public boolean isValid(final DockstoreYaml12 value, final ConstraintValidatorContext context) {
-        return value.getService() != null || !value.getWorkflows().isEmpty() || !value.getTools().isEmpty();
+        return value.getService() != null
+            || value.getWorkflows() != null && !value.getWorkflows().isEmpty()
```
[the remark this answers is not in the document]
Comment (svonworl) [reply dockstore-common/src/main/java/io/dockstore/common/yaml/Dockstore12Validator.java:33]: It was failing when the `workflows` property was set to nothing (ala `workflows:` with no associated yaml).  I believe the class validator was, for whatever reason, being called before the `NotNull`, and didn't check for null, so it died.

**Review by Richard-Hansen — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java [hunk 2]
```diff
@@ -391,6 +391,14 @@ protected void githubWebhookRelease(String repository, String username, String g
         }
     }
 
+    private String createMessageFromException(Exception ex) {
+        String message = ex instanceof CustomWebApplicationException ? ((CustomWebApplicationException)ex).getErrorMessage() : ex.getMessage();
+        if (ex instanceof DockstoreYamlHelper.DockstoreYamlException) {
+            message = "Error reading .dockstore.yml: " + message;
+        }
```
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:398]: DockstoreYamlHelper already has a constant for `Error reading .dockstore.yml`, so could re-use that here

**Review by kathy-t — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]
All three Codacy code smells are simple and worth doing
Otherwise, thanks for reviewing the tests, looks good

**Comment by svonworl** [comment]
> All three Codacy code smells are simple and worth doing Otherwise, thanks for reviewing the tests, looks good

Regarding the Sonarcloud feedback:

SonarCloud is mistaken about the two "can't be null" recommendations, the enclosing class is invoked by the validator before the `NotNull` constraints are checked on the properties in question.  They can actually be null when the code is invoked, if `workflows:` is set to null in the yaml.

Re the other item, when I wrote the code, I pondered `v -> buildMessageFromViolation(v)` versus `DockstoreYamlHelper::buildMessageFromViolation`, and I went with the former, because it's shorter and immediately apparent that it's calling a method in the same class.  If something like `class::buildMessageFromViolation` was legal java, I probably would have preferred that.

**Comment by denis-yuen** [comment]
> Regarding the Sonarcloud feedback:

Seems reasonable, I think you can let SonarCloud know via annotation 
https://www.baeldung.com/sonar-exclude-violations
Might also let future us know since there's a chance they will be accidentally and automatically factored out via wizard

**Review by coverbeck — APPROVED** [review]

**Pull request merged**